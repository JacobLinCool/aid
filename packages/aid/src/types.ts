import type { OpenAI } from "openai";
import type { ZodIssue, ZodSchema } from "zod";

export type PromiseOr<T> = T | Promise<T>;

export type QueryEngine<QueryPayload> = (payload: QueryPayload) => PromiseOr<string>;

export type FormatEngine<TaskPayload, QueryPayload> = (
	payload: TaskPayload,
	schema: ZodSchema,
) => PromiseOr<QueryPayload>;

export type TaskBuilder<TaskGoal, CaseParam, TaskPayload> = (
	goal: TaskGoal,
	param: CaseParam | undefined,
	examples: [CaseParam, unknown][],
) => PromiseOr<TaskPayload>;

export type BaseChatParam =
	| string
	| {
			text: string;
	  };

export type VisionChatParam =
	| string
	| {
			text?: string;
			images?: OpenAI.Chat.ChatCompletionContentPartImage.ImageURL[];
	  };

/**
 * Options for a custom Aid Task.
 */
export interface AidTaskOptions<In, Out> {
	/**
	 * The few-shot prompt examples.
	 */
	examples?: [In, Out][];

	/**
	 * Whether to check the output against the schema.
	 * @default false
	 */
	check?: boolean;

	/**
	 * The default input (last user message) if no input is provided.
	 */
	default?: In;
}

export type AidTaskRunner<In, Out> = (input?: In) => Promise<{ result: Out; errors: ZodIssue[] }>;

export type BaseChatMessageRole = "system" | "user" | "assistant";

export type BaseChatMessage<R extends BaseChatMessageRole = BaseChatMessageRole> = {
	role: R;
	content: string;
};
