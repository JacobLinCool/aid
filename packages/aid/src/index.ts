export * from "./aid";
export * from "./types";

export * from "./query/cohere";
export * from "./query/openai";

export * from "./format/json";

export * from "./task/chat";
export * from "./task/vision-chat";
