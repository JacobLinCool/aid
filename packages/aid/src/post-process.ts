// post process the output from the large language model, and make it parseable by the JSON parser
export function post_process(llm_output: string): string {
	llm_output = llm_output.trim();

	// remove possible markdown code fence at the beginning and end
	llm_output = llm_output.replace(/^```.*\n/m, "").replace(/```\n?$/m, "");

	return llm_output;
}
