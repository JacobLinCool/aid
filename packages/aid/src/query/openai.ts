import debug from "debug";
import type { OpenAI } from "openai";
import type { BaseChatMessage, QueryEngine } from "../types";

const log = debug("aid:openai");

export const OpenAIQuery = (
	openai: OpenAI,
	param: Omit<OpenAI.Chat.ChatCompletionCreateParamsNonStreaming, "messages">,
): QueryEngine<BaseChatMessage[] | OpenAI.Chat.ChatCompletionMessageParam[]> => {
	const q: QueryEngine<BaseChatMessage[] | OpenAI.Chat.ChatCompletionMessageParam[]> = async (
		messages,
	) => {
		const payload: OpenAI.Chat.ChatCompletionCreateParamsNonStreaming = {
			temperature: 0,
			...param,
			messages,
		};

		log("payload", payload);
		const res = await openai.chat.completions.create(payload);
		log({
			id: res.id,
			fingerprint: res.system_fingerprint,
			usage: res.usage,
			reply: res.choices[0],
		});

		const text = res.choices[0].message.content;
		if (!text) {
			throw new Error("No text returned from OpenAI");
		}
		log("text", text);

		return text;
	};
	return q;
};
