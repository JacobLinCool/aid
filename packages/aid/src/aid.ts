import debug from "debug";
import type { OpenAI } from "openai";
import type { z } from "zod";
import { DefaultJSON, VisionJSON } from "./format/json";
import { post_process } from "./post-process";
import { OpenAIQuery } from "./query/openai";
import { ChatTask } from "./task/chat";
import { VisionChatTask } from "./task/vision-chat";
import type {
	AidTaskOptions,
	AidTaskRunner,
	BaseChatMessage,
	BaseChatParam,
	FormatEngine,
	QueryEngine,
	TaskBuilder,
	VisionChatParam,
} from "./types";

const log = debug("aid");

export class Aid<TaskGoal, CaseParam, FormatPayload, QueryPayload> {
	protected qe: QueryEngine<QueryPayload>;
	protected fe: FormatEngine<FormatPayload, QueryPayload>;
	protected tb: TaskBuilder<TaskGoal, CaseParam, FormatPayload>;

	constructor(
		qe: QueryEngine<QueryPayload>,
		fe: FormatEngine<FormatPayload, QueryPayload>,
		tb: TaskBuilder<TaskGoal, CaseParam, FormatPayload>,
	) {
		this.qe = qe;
		this.fe = fe;
		this.tb = tb;
	}

	task<Out>(
		goal: TaskGoal,
		expected: z.ZodType<Out>,
		opt?: AidTaskOptions<CaseParam, Out>,
	): AidTaskRunner<CaseParam, Out> {
		return async (input = opt?.default) => {
			const fp = await this.tb(goal, input, opt?.examples ?? []);
			log("fp", fp);

			const qp = await this.fe(fp, expected);
			log("qp", qp);

			const output = await this.qe(qp);
			log("output", output);

			const post = post_process(output);
			log("post-processed", post);

			const json = JSON.parse(post);

			if (!opt?.check) {
				return { result: json, errors: [] };
			}

			const result = expected.safeParse(json);
			if (result.success) {
				return { result: json, errors: [] };
			} else {
				return { result: json, errors: result.error.issues };
			}
		};
	}

	static from(
		openai: OpenAI,
		param: Omit<OpenAI.Chat.ChatCompletionCreateParamsNonStreaming, "messages">,
	): Aid<string, BaseChatParam, BaseChatMessage[], BaseChatMessage[]> {
		const qe = OpenAIQuery(openai, param);
		const fe = DefaultJSON({ strategy: "json-schema" });
		const tb = ChatTask();
		return new Aid(qe, fe, tb);
	}

	static chat(
		qe: QueryEngine<BaseChatMessage[]>,
		{
			tb = ChatTask(),
			fe = DefaultJSON({ strategy: "json-schema" }),
		}: {
			tb?: TaskBuilder<string, BaseChatParam, BaseChatMessage[]>;
			fe?: FormatEngine<BaseChatMessage[], BaseChatMessage[]>;
		} = {},
	): Aid<string, BaseChatParam, BaseChatMessage[], BaseChatMessage[]> {
		return new Aid(qe, fe, tb);
	}

	static vision(
		qe: QueryEngine<OpenAI.Chat.ChatCompletionMessageParam[]>,
		{
			tb = VisionChatTask(),
			fe = VisionJSON({ strategy: "json-schema" }),
		}: {
			tb?: TaskBuilder<string, VisionChatParam, OpenAI.Chat.ChatCompletionMessageParam[]>;
			fe?: FormatEngine<
				OpenAI.Chat.ChatCompletionMessageParam[],
				OpenAI.Chat.ChatCompletionMessageParam[]
			>;
		} = {},
	): Aid<
		string,
		VisionChatParam,
		OpenAI.Chat.ChatCompletionMessageParam[],
		OpenAI.Chat.ChatCompletionMessageParam[]
	> {
		return new Aid(qe, fe, tb);
	}
}
