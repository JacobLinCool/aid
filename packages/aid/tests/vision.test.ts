import debug from "debug";
import fs from "node:fs";
import { OpenAI } from "openai";
import { describe, expect, it } from "vitest";
import { z } from "zod";
import { Aid, OpenAIQuery } from "../src";

debug.enable("aid*");

describe("Vision", () => {
	it(
		"Vision Task",
		async (c) => {
			const api_key = process.env.OPENAI_API_KEY;
			if (!api_key) {
				c.skip();
				return;
			}

			const openai = new OpenAI({ apiKey: api_key });
			const aid = Aid.vision(
				OpenAIQuery(openai, { model: "gpt-4-vision-preview", max_tokens: 2048 }),
			);

			const analyze = aid.task(
				"Analyze the person in the image",
				z.object({
					gender: z.enum(["boy", "girl", "other"]),
					age: z.enum(["child", "teen", "adult", "elderly"]),
					emotion: z.enum(["happy", "sad", "angry", "surprised", "neutral"]),
					clothing: z.string().max(100),
					background: z.string().max(100),
				}),
			);

			const datauri = `data:image/png;base64,${fs.readFileSync(
				"packages/aid/tests/fixtures/dall-e.png",
				"base64",
			)}`;

			const { result } = await analyze({ images: [{ url: datauri }] });
			expect(typeof result).toBe("object");
			// expect(result).toMatchSnapshot();
		},
		{ timeout: 60_000 },
	);
});
