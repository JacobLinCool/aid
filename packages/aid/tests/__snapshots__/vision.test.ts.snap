// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vision > Vision Task 1`] = `
{
  "age": "teen",
  "background": "plain white",
  "clothing": "black shirt with a brown tie",
  "emotion": "neutral",
  "gender": "boy",
}
`;
